contracts/ERC1155M.sol

```diff
@@ -379,9 +415,18 @@ contract ERC1155M is IERC1155M, ERC1155Supply, ERC2981, Ownable2Step, Reentrancy
         uint256 tokenId,
         uint32 qty,
         uint32 limit,
-        bytes32[] calldata proof
+        bytes32[] calldata proof,
+        uint64 timestamp,
+        bytes memory signature
     ) internal hasSupply(tokenId, qty) {
         uint64 stageTimestamp = uint64(block.timestamp);
+
+        if (_cosigner != address(0)) {
+            assertValidCosign(msg.sender, tokenId, qty, timestamp, signature);
+            _assertValidTimestamp(timestamp);
+            stageTimestamp = timestamp;
```

Comment:
Because of this line, wouldn't it be possible for the cosigner to rewind the minting stages? For example, the only thing validating the timestamp in `_assertValidTimestamp` is that the timestamp value in the payload is < the current block timestamp. So after one stage has passed, would this value be game-able to mint after the minting window has passed?

Reply:
Yeah good call. This behavior is intentional. We want to use payload time (guarded by signature) as "mint time". Previously, we have seen many users failed to mint. When they grab signature, the stage is still valid, but when they submit tx, stage is ended. The user experience for them is very bad so we decided to use they timestamp when they grab signature. It's more like extend extra few minutes for those who intent to mint. We still have other guards such as whitelist, supply etc.